# Coding Exercises (Part 2)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Basics

### Exercise 2: Indexing and Slicing

7. -__Import Pandas__ (as pd)<br>
-__Import__ the Listings Dataset (__listings.csv__) and save the DataFrame in the variable __listings__ and<br>
-__Inspect__ the __first 5 Rows__!

In [1]:
import pandas as pd

In [2]:
listings = pd.read_csv("listings.csv")

In [3]:
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

8. __Select__ the column __"Name"__ with the __attribute__ notation! What is the __company name__ with __index label 4__?

In [4]:
listings.Name

0                              Agilent Technologies, Inc.
1                                       Alcoa Corporation
2                                             Altaba Inc.
3                                      AAC Holdings, Inc.
4                           American Airlines Group, Inc.
5                        Altisource Asset Management Corp
6                           Atlantic American Corporation
7                                      Aaron&#39;s,  Inc.
8                           Applied Optoelectronics, Inc.
9                                              AAON, Inc.
10                                 Advance Auto Parts Inc
11                                             Apple Inc.
12                            American Assets Trust, Inc.
13                                 Almaden Minerals, Ltd.
14                           Atlas Air Worldwide Holdings
15      iShares MSCI All Country Asia ex Japan Index Fund
16                                  Axon Enterprise, Inc.
17            

The company name is ... American Airlines Group, Inc.!

9. Apply also the __second alternative__ to select the column __"Name"__!

In [5]:
listings["Name"]

0                              Agilent Technologies, Inc.
1                                       Alcoa Corporation
2                                             Altaba Inc.
3                                      AAC Holdings, Inc.
4                           American Airlines Group, Inc.
5                        Altisource Asset Management Corp
6                           Atlantic American Corporation
7                                      Aaron&#39;s,  Inc.
8                           Applied Optoelectronics, Inc.
9                                              AAON, Inc.
10                                 Advance Auto Parts Inc
11                                             Apple Inc.
12                            American Assets Trust, Inc.
13                                 Almaden Minerals, Ltd.
14                           Atlas Air Worldwide Holdings
15      iShares MSCI All Country Asia ex Japan Index Fund
16                                  Axon Enterprise, Inc.
17            

10. __Select__ the listing/row at __index position 100__ with the __iloc__ operator (all columns)! What is the IPO_Year?

In [6]:
listings.iloc[100]

Symbol                                                     ADVM
Exchange                                                 NASDAQ
Name                              Adverum Biotechnologies, Inc.
Last_Price                                                 5.42
Market_Cap                                          3.42549e+08
ADR TSO                                                     NaN
IPO_Year                                                   2014
Sector                                              Health Care
Industry      Biotechnology: Biological Products (No Diagnos...
Name: 100, dtype: object

The IPO_Year of this listing is ... 2014!

11. __Select__ the listing/row at __index position 2000__ with the __iloc__ operator (and only the column __"Name"__)! <br>
What is the __company name__ of this listing?

In [7]:
listings.iloc[2000, 2]

'Eagle Bancorp, Inc.'

It´s the ... Eagle Bancorp, Inc.!

12. __Select__ the __last 10 rows__ and the columns __Name, Last_Price, Sector__ and __Industry__ with the __iloc__ operator! __Fill in the Gaps__! <br>
What is the __Sector__ of the __Zymeworks Inc.__?

In [8]:
listings.iloc[-10:, [2, 3, 7, 8]]

Name  Last_Price  \
6842                  Zosano Pharma Corporation        3.71   
6843                          BATS BZX Exchange       10.10   
6844                  ZTO Express (Cayman) Inc.       19.11   
6845  Virtus Global Dividend & Income Fund Inc.       10.86   
6846                                Zoetis Inc.      101.33   
6847                                Zumiez Inc.       26.72   
6848                                Zuora, Inc.       19.79   
6849                             Zymeworks Inc.       15.74   
6850              Zynerba Pharmaceuticals, Inc.        7.85   
6851                                Zynex, Inc.        5.02   

                 Sector                                           Industry  
6842        Health Care                              Major Pharmaceuticals  
6843                NaN                                                NaN  
6844     Transportation                  Trucking Freight/Courier Services  
6845                NaN                                                NaN  
6846        Health Care                              Major Pharmaceuticals  
6847  Consumer Services                     Clothing/Shoe/Accessory Stores  
6848         Technology            Computer Software: Prepackaged Software  
6849        Health Care                              Major Pharmaceuticals  
6850        Health Care                              Major Pharmaceuticals  
6851        Health Care  Biotechnology: Electromedical & Electrotherape...

The Sector is ... Health Care!

13. __Import__ the __Listings Dataset__ one more time (save in the variable __listings__) and set the column __"Symbol"__ as the __index__! <br>
__Fill in the Gaps__!

In [9]:
listings = pd.read_csv("listings.csv", index_col = "Symbol")

In [10]:
#run the cell and inspect!
listings.head()

Exchange                           Name  Last_Price    Market_Cap  \
Symbol                                                                     
A          NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
AA         NYSE              Alcoa Corporation       29.15  5.407810e+09   
AABA     NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
AAC        NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
AAL      NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

        ADR TSO  IPO_Year            Sector  \
Symbol                                        
A           NaN    1999.0     Capital Goods   
AA          NaN    2016.0  Basic Industries   
AABA        NaN       NaN        Technology   
AAC         NaN    2014.0       Health Care   
AAL         NaN       NaN    Transportation   

                                                Industry  
Symbol                                                    
A       Biotechnology: Laboratory Analytical Instruments  
AA                                              Aluminum  
AABA                                        EDP Services  
AAC                                 Medical Specialities  
AAL                        Air Freight/Delivery Services

In [11]:
#run the cell and inspect!
listings.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6852 entries, A to ZYXI
Data columns (total 8 columns):
Exchange      6852 non-null object
Name          6852 non-null object
Last_Price    6745 non-null float64
Market_Cap    5954 non-null float64
ADR TSO       140 non-null float64
IPO_Year      3105 non-null float64
Sector        5309 non-null object
Industry      5309 non-null object
dtypes: float64(4), object(4)
memory usage: 481.8+ KB


14. __Inspect__ the new Index of the DataFrame by calling the __index__ attribute!

In [12]:
listings.index

Index(['A', 'AA', 'AABA', 'AAC', 'AAL', 'AAMC', 'AAME', 'AAN', 'AAOI', 'AAON',
       ...
       'ZSAN', 'ZTEST', 'ZTO', 'ZTR', 'ZTS', 'ZUMZ', 'ZUO', 'ZYME', 'ZYNE',
       'ZYXI'],
      dtype='object', name='Symbol', length=6852)

15. __Select__ the __row__ with the index label __"AAPL"__ (Apple) (use the __loc__ operator)! What is the __Last_Price__?

In [13]:
listings.loc["AAPL"]

Exchange                      NASDAQ
Name                      Apple Inc.
Last_Price                    200.62
Market_Cap               9.45979e+11
ADR TSO                          NaN
IPO_Year                        1980
Sector                    Technology
Industry      Computer Manufacturing
Name: AAPL, dtype: object

The Last_Price is ... 200.62!

16. __Select__ the __row__ with the index label __"FB"__ (Facebook) and only the columns __"IPO_Year"__ and __"Industry"__! <br>
What´s the __Industry__?

In [14]:
listings.loc["FB", ["IPO_Year", "Industry"]]

IPO_Year                                               2012
Industry    Computer Software: Programming, Data Processing
Name: FB, dtype: object

The Industry is ... Computer Software: Programming, Data Processing!

17. __Select__ the columns __"Name"__ and __"Last_Price"__ with the __loc__ operator (all rows)!

In [15]:
listings.loc[:, ["Name", "Last_Price"]]

Name  Last_Price
Symbol                                                               
A                              Agilent Technologies, Inc.     81.6800
AA                                      Alcoa Corporation     29.1500
AABA                                          Altaba Inc.     75.3900
AAC                                    AAC Holdings, Inc.      2.1600
AAL                         American Airlines Group, Inc.     34.0200
AAMC                     Altisource Asset Management Corp     29.9000
AAME                        Atlantic American Corporation      2.4800
AAN                                    Aaron&#39;s,  Inc.     53.5400
AAOI                        Applied Optoelectronics, Inc.     12.3000
AAON                                           AAON, Inc.     44.8800
AAP                                Advance Auto Parts Inc    178.8000
AAPL                                           Apple Inc.    200.6200
AAT                           American Assets Trust, Inc.     46.0800
AAU                                Almaden Minerals, Ltd.      0.5750
AAWW                         Atlas Air Worldwide Holdings     49.4900
AAXJ    iShares MSCI All Country Asia ex Japan Index Fund     73.1400
AAXN                                Axon Enterprise, Inc.     60.1200
AB                         AllianceBernstein Holding L.P.     29.3600
ABB                                               ABB Ltd     19.4400
ABBV                                          AbbVie Inc.     82.9500
ABC            AmerisourceBergen Corporation (Holding Co)     74.0000
ABCB                                       Ameris Bancorp     35.0900
ABDC                               Alcentra Capital Corp.      7.8100
ABEO                             Abeona Therapeutics Inc.      7.5400
ABEOW                            Abeona Therapeutics Inc.      3.0700
ABEV                                           Ambev S.A.      4.5400
ABG                           Asbury Automotive Group Inc     72.6400
ABIL                                         Ability Inc.      1.7000
ABIO                                 ARCA biopharma, Inc.      5.9200
ABM                           ABM Industries Incorporated     37.0100
...                                                   ...         ...
ZEAL                                   Zealand Pharma A/S     19.0100
ZEN                                         Zendesk, Inc.     83.3400
ZEUS                                  Olympic Steel, Inc.     17.3900
ZF                          Virtus Total Return Fund Inc.      9.8500
ZFGN                                         Zafgen, Inc.      2.7300
ZG                                     Zillow Group, Inc.     36.4300
ZGNX                                        Zogenix, Inc.     40.0000
ZION                            Zions Bancorporation N.A.     47.1400
ZIONW                           Zions Bancorporation N.A.     14.7400
ZIOP                                ZIOPHARM Oncology Inc      4.3300
ZIV                                      Credit Suisse AG     75.1200
ZIXI                                      Zix Corporation      7.3400
ZKIN                      ZK International Group Co., Ltd      1.8100
ZLAB                                      Zai Lab Limited     32.4300
ZN                                     Zion Oil & Gas Inc      0.8188
ZNGA                                           Zynga Inc.      5.4700
ZNH               China Southern Airlines Company Limited     48.7400
ZNWAA                                  Zion Oil & Gas Inc         NaN
ZOM                        Zomedica Pharmaceuticals Corp.      0.3300
ZS                                          Zscaler, Inc.     68.4400
ZSAN                            Zosano Pharma Corporation      3.7100
ZTEST                                   BATS BZX Exchange     10.1000
ZTO                             ZTO Express (Cayman) Inc.     19.1100
ZTR             Virtus Global Dividend & Income Fund Inc.     10.8600
ZTS                                           Zoetis Inc.    10

18. __Select__ the last three columns __"IPO_Year"__, __"Sector"__ and __"Industry"__ for __"MSFT"__ (Microsoft) and __"DIS"__ (Disney) with the __loc__ operator. Don´t use "Sector" or "Industry" in your code! What is Disney´s __Industry__?

In [16]:
listings.loc[["MSFT", "DIS"], "IPO_Year":]

IPO_Year             Sector                                 Industry
Symbol                                                                      
MSFT      1986.0         Technology  Computer Software: Prepackaged Software
DIS          NaN  Consumer Services                      Television Services

Disney´s Industry is ... Television Services

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Hints (Spoiler!)

7. pandas as pd, pd.read_csv("filename"), head()

8. listings.---

9. listings["---"]

10. listings.iloc[---]

11. listings.iloc[row index position, column index position]; "Name" column is at index position ... 2

12. The last 10 rows = from index position -10 till the end (:); Count the column index positions of the Columns


13. parameter index_col

14. listings.---

15. listings.loc["row label"]

16. listings.loc[---, [---, ---]]

17.  listings.loc[ : , [---, ---]]

18. listings.loc[[---, ---], ---:]